### DISCUSSION TITLE
Facing Issue while Planning Eks Cluster, How to By Pass the error ?

### DISCUSSION THREAD

**Discussion by vramanufidato** [gruntwork-io/knowledge-base#644] [How do I]
Currently we are doing the validate and Plan for eks cluster. Getting the below error.

Error
**│ Error: Failed to configure client: exec plugin: invalid apiVersion "client.authentication.k8s.io/v1alpha1"
│   with module.eks_k8s_role_mapping.kubernetes_config_map.eks_to_k8s_role_mapping,
   on .terraform/modules/eks_k8s_role_mapping/modules/eks-k8s-role-mapping/main.tf line 33, in resource "kubernetes_config_map" "eks_to_k8s_role_mapping":
33: resource "kubernetes_config_map" "eks_to_k8s_role_mapping" {**

in _envcommon seeing below segment 

Code : **_envcommon/services/eks-cluster.hcl line 151**
iam_role_to_rbac_group_mapping = {
    "arn:aws:iam::${local.common_vars.locals.account_ids[local.account_name]}:role/GruntworkAccountAccessRole" = ["system:masters"]
    (dependency.ops_admin_role.outputs.iam_role_arn)                                                           = ["system:masters"]
    (dependency.baseline.outputs.allow_dev_access_from_other_accounts_iam_role_arn)     = ["system:masters"]
    (dependency.ecs_deploy_runner.outputs.ecs_task_iam_roles["terraform-planner"]["arn"])   = ["system:masters"]
    **(dependency.ecs_deploy_runner.outputs.ecs_task_iam_roles["terraform-applier"]["arn"])     = ["system:masters"]**
  }

As we are using Plan only currently Not sure why this error is coming up ? Is there any other way to by pass and do the plan ?

---

<ins datetime="2023-01-25T11:36:17Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109834">Tracked in ticket #109834</a></p>
</ins>

**Comment by ryehowell** [comment]
Hi @bangupoc. Could you provide what version of the service catalog/EKS module(s) you are using as well as what version of Kuberntes/EKS you are specifying?

**Reply by vramanufidato** [reply]
Hi @ryehowell i am seeing 

terraform {
  source = "${local.source_base_url}?ref=v0.95.0"
}

Apart from that 

source_base_url = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/services/eks-cluster"

we have specified eks-workers-v0.95.0-*

**Reply by ryehowell** [reply]
Hi @bangupoc can you do a search for `kubernetes_version` and report back what version is being set for that (if any)?

**Reply by vramanufidato** [reply]
Hi @ryehowell its having kubernetes_version "1.13"

**Reply by ryehowell** [reply]
Hi @bangupoc, kubernetes version `1.13` is an unsupported version on AWS EKS. I would suggest setting `kubernetes_version` to a more recent and supported version based on the [AWS EKS supported versions](https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html#available-versions) and ensure that it is a version that is supported by the `v0.95.0` release of the `terraform-aws-service-catalog` (you could use the default version of `1.22` as a safe supported version). The reason for the error is most likely due to kubernetes version `1.13` not having support for the `client.authentication.k8s.io/v1alpha1` api version which is causing the error during the planning stage.

**Reply by vramanufidato** [reply]
Hi @ryehowell i am not seeing any such reference in infrastructure-live code base, the changes are in the modules only so we need to update the modules in case required ? It is referring to the latest modules of GW, as per code i am seeing the service-catalog is using version 1.24 when i downloaded the latest service catalog as the envcommon  source_base_url points to GW repo. As per my understanding it should take 1.24 version only. As of now we are not maintaining separate module repo we are pointing to GW modules.
source_base_url = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/services/eks-cluster"

**Reply by ryehowell** [reply]
Hi @bangupoc, I'm assuming your `infrastructure-live` repo/codebase is the codebase you are maintaining, correct? Based on what you provided it looks like you are consuming the `terraform-aws-service-catalog.git//modules/services/eks-cluster`, version tag `v0.95.0`. You mentioned that also in the `infrastructure-live` repo that the input variable `kubernetes_version` is being set to the value of `1.13`, is this correct? If so, then that value will override the default value that is set in the underlying module (which it sounds like you downloaded to view). If these assumptions are correct, then you should be able to get past the error you are getting by setting the `kubernetes_version` to `1.22` in your `infrastructure-live` repo which is the same version that is defaulted in the `v0.95.0` release of the service catalog. 

Essentially any input variables that you define in your `infrastructure-live` repo will override the "default" values that are set in the underlying modules. To reference what the default values are for a given version/release, you can browse the underlying module (as you mentioned doing by downloading) as a reference point. 

A working example can be found here: https://github.com/gruntwork-io/terraform-aws-service-catalog/tree/master/examples/for-production/infrastructure-live

**Reply by vramanufidato** [reply]
Hi @ryehowell Thanks for quick response, but i am not seeing kubernetes_version is been passed from infrastructure-live only in modules these variables are been used as i mentioned we are not using any separate module repo and its pointing to source_base_url = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/services/eks-cluster" doesn't it mean it is already taking the kubernetes_version = 1.24. The kubernetes_version = 1.3 which i mentioned was the modeule code  which i downloaded few months back using it as a reference to develop the code, it was a mistake from my side that i mentioned 1.3 by refering the code which i have been referring but actually it is 1.24 as i explained. But how often the code changes happens is there a way GW updates the customer in the site whenever a new module change happens?

**Reply by ryehowell** [reply]
Hi @bangupoc, I see now 😃 and this makes more sense. Since you are using kubernetes `1.24`, you will need to update the service catalog to the latest release which is `v0.100.0` as kubernetes/eks `1.24` has multiple backward incompatible changes that need to be accounted for. One of the breaking changes that was introduced was the removal of the `client.authentication.k8s.io/v1alpha1` api which is shown in the initial error you provided. The new api version for that is `client.authentication.k8s.io/v1` which is reflected in the latest version of the service catalog. 

Please take a look at the latest [service catalog v0.100.0 release notes](https://github.com/gruntwork-io/terraform-aws-service-catalog/releases/tag/v0.100.0) of the service catalog for detailed information on this.

**Reply by ryehowell** [reply]
Hi @bangupoc! I just wanted to follow up to see if you were able to resolve the issue you were facing based on the feedback provided. If so, please feel free to mark the question as answered so any users in the future that run into the same issue can have a KB to help them. If not, please follow up with any new information so we can help get the issue resolved. Cheers!

**Reply by vramanufidato** [reply]
Hi @ryehowell as we are using the GW modules and service catalog still we are facing the issue. We are still troubleshooting the issue will keep you posted. There are some forums mentioned we need awsv2 version so we are checking from that perspective as well. Thanks for support @ryehowell i appreciate your guidance in this issue.

**Reply by ryehowell** [reply]
Hi @bangupoc, thanks for following up. I actually think we found the issue that you are running in to. This is an issue that is most likely being caused by a change introduced in the upstream Terraform Kubernetes provider which silently included some breaking changes in 2.17 (removal of the "client.authentication.k8s.io/v1alpha1") during a dep update to their Go packages. This was not reflected in the latest Release Notes and has caused some headaches in the broader community. 

If you can confirm that during your plan run that you are pulling in: `hashicorp/kubernetes v2.17.0` then the workaround is to add an extra version constraint to the `hashicorp/kubernetes` provider to `version = "< 2.17"` and it should fix the issue you are facing. Depending on your code setup this could be an extra tf file in your root module or if you are using terragrunt then you can place it as a `generate` block in one of your common files. 

Do want to give that a go and follow up?

**Reply by ryehowell** [reply]
Hi @bangupoc,

I just wanted to follow up to see if the provided workaround resolved the issue you were facing. Were you able to resolve the issue?

**Reply by vramanufidato** [reply]
Hi @ryehowell still the error is not resolved tried using the ?ref=v0.100.0 and pined the kubernetes version to 2.17 as well still v1aplha1 error is coming up and its mentioned https://www-app.gruntwork.io/repos/v0.14.0/terraform-aws-eks/modules/eks-k8s-role-mapping/README.md#aws-auth-configmap-generator-binary the iam roles will not be created and they have provide kubectl ref any idea about it what needs to be done for it ?

**Reply by gorkemgoknar** [reply]
I am facing a similar problem (trying to get from v0.82.0 to v0.100.0 before AWS updates 1.21 to 1.22)
and did override so uses helm/kubernetes <2.17 

Override in reference architecture does not seem to override this one : https://github.com/gruntwork-io/terraform-aws-eks/blob/94aef4221985bfd314ef33aaf0f956f035cadfa3/modules/eks-k8s-role-mapping/main.tf#L21
```

getting credentials: exec plugin is configured to use API version client.authentication.k8s.io/v1, plugin returned version client.authentication.k8s.io/v1alpha1
[ecs-deploy-runner][2023-02-18T17:52:05+0000] │ 
[ecs-deploy-runner][2023-02-18T17:52:05+0000] │   with module.eks_k8s_role_mapping.kubernetes_config_map.eks_to_k8s_role_mapping,
[ecs-deploy-runner][2023-02-18T17:52:05+0000] │   on .terraform/modules/eks_k8s_role_mapping/modules/eks-k8s-role-mapping/main.tf line 36, in resource "kubernetes_config_map" "eks_to_k8s_role_mapping":
[ecs-deploy-runner][2023-02-18T17:52:05+0000] │   36: resource "kubernetes_config_map" "eks_to_k8s_role_mapping" {
[ecs-deploy-runner][2023-02-18T17:52:05+0000] │ 
[ecs-deploy-runner][2023-02-18T17:52:05+0000] ╵
[ecs-deploy-runner][2023-02-18T17:52:05+0000] time=2023-02-18T17:52:05Z level=error msg=1 error occurred:
[ecs-deploy-runner][2023-02-18T17:52:05+0000] 	* exit status 1
```

**Reply by gorkemgoknar** [reply]
At last found the problem after a day 
for a v1alpha1 supported api EKS (1.21,1.22,1.23) The final version that does not cause trouble is terraform-aws-service-catalog -> v0.99.2 . on v0.100.0  eks-cluster uses its own provider and uses v1 api (terraform override has no use). on v0.99.2 it uses v1alpha1 api.  You need a provider kubernetes override withing eks cluster you have to use your own version, or use v0.99.2 (for supported kubernetes releases <=1.23 I think) 

Note that for v0.100.0 you will have problems with other dependend modules that use rest api for > v0.100.0 service catalog or terraform-aws-eks.
You should still use on root provider override hashicorp/kubernetes v2.17.0 as 2.18 and onwards use v1. (it will give trouble with namespace etc and other modules if you do not).

**Reply by ryehowell** [reply]
Hi everyone, just for future reference, a KB article was created to address this issue. Please see the [article](https://github.com/gruntwork-io/knowledge-base/discussions/671) for details. Hope this helps!
